File: src/main/java/it/pagopa/ecommerce/utils/ApiKeyFilter.java
package it.pagopa.ecommerce.utils;

import io.quarkus.vertx.web.RouteFilter;
import io.vertx.ext.web.RoutingContext;
import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.jboss.logging.Logger;

import java.util.List;
import java.util.Set;

@ApplicationScoped
public class ApiKeyFilter {

    private static final Logger LOG = Logger.getLogger(ApiKeyFilter.class);

    @ConfigProperty(name = "security.apiKey.primary")
    String primaryApiKey;

    @ConfigProperty(name = "security.apiKey.secondary")
    String secondaryApiKey;

    @ConfigProperty(name = "security.apiKey.securedPaths")
    List<String> securedPaths;

    private Set<String> getValidApiKeys() {
        return Set.of(primaryApiKey, secondaryApiKey);
    }

    @RouteFilter
    void filter(RoutingContext ctx) {
        String path = ctx.request().path();

        if (securedPaths.stream().anyMatch(path::startsWith)) {
            String apiKey = ctx.request().getHeader("x-api-key");

            if (!isValidApiKey(apiKey)) {
                LOG.errorf("Unauthorized request for path %s - Missing or invalid API key", path);
                ctx.response().setStatusCode(401).end();
                return;
            }

            logWhichApiKey(apiKey, path);
        }

        ctx.next();
    }

    private boolean isValidApiKey(String apiKey) {
        return apiKey != null && !apiKey.trim().isEmpty() && getValidApiKeys().contains(apiKey);
    }

    private void logWhichApiKey(String apiKey, String path) {
        String apiKeyType;
        if (primaryApiKey.equals(apiKey)) {
            apiKeyType = "primary";
        } else if (secondaryApiKey.equals(apiKey)) {
            apiKeyType = "secondary";
        } else {
            apiKeyType = "unknown";
        }
        LOG.debugf("API key type used for path %s: %s", path, apiKeyType);
    }
}

File: src/main/kotlin/utils/ApiKeyFilter.kt
package utils

import io.quarkus.vertx.web.RouteFilter
import io.vertx.ext.web.RoutingContext
import jakarta.enterprise.context.ApplicationScoped
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.jboss.logging.Logger

@ApplicationScoped
class ApiKeyFilter {

    companion object {
        private val LOG: Logger = Logger.getLogger(ApiKeyFilter::class.java)
    }

    @ConfigProperty(name = "security.apiKey.primary") lateinit var primaryApiKey: String

    @ConfigProperty(name = "security.apiKey.secondary") lateinit var secondaryApiKey: String

    @ConfigProperty(name = "security.apiKey.securedPaths") lateinit var securedPaths: List<String>

    private val validApiKeys: Set<String>
        get() = setOf(primaryApiKey, secondaryApiKey)

    @RouteFilter
    fun filter(ctx: RoutingContext) {
        val path = ctx.request().path()

        if (securedPaths.any { path.startsWith(it) }) {
            val apiKey = ctx.request().getHeader("x-api-key")

            if (!isValidApiKey(apiKey)) {
                LOG.errorf("Unauthorized request for path %s - Missing or invalid API key", path)
                ctx.response().setStatusCode(401).end()
                return
            }

            logWhichApiKey(apiKey, path)
        }

        ctx.next()
    }

    private fun isValidApiKey(apiKey: String?): Boolean {
        return !apiKey.isNullOrBlank() && validApiKeys.contains(apiKey)
    }

    private fun logWhichApiKey(apiKey: String?, path: String) {
        val apiKeyType =
            when (apiKey) {
                primaryApiKey -> "primary"
                secondaryApiKey -> "secondary"
                else -> "unknown"
            }
        LOG.debugf("API key type used for path %s: %s", path, apiKeyType)
    }
}
